metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala

```diff
@@ -556,8 +556,8 @@ object BuildTargets {
                 // in Bloop to compile only `B` than `A+B`.
                 leaves += head
             }
-            loop(tail)
           }
+          loop(tail)
```

Comment:
4h of debugging and narrowing bug source only to change 1 line :D

Reply:
Wow, wow, great catch!
So, previously, we stopped looping when the `head` is already visited. This is a problem.

Consider the following dependencies (`B -> A` represents "B depends on A").

```mermaid
graph TD
    A --> B
    A --> C
    B --> C
    B --> D
```

When we `inverseDependencies(A, graph)`, it **sometimes** ignored `D` (depends on the search order).

For example, it ignore `D` in the following scenario

- `loop([A]), isVisited = {}`
- `loop([C,B]), isVisited = {A}`
- `loop([B]), isVisited = {A, C}`
- `loop([C,D]), isVisited={A,C,B}` stop looping here because `C` is already visited! even though we haven't visited D yet!!!

but it visits all members in the following scenario

- `loop([A]), isVisited = {}`
- `loop([B, C]), isVisited = {A}` note that we start from B
  - `loop([C,D]), isVisited = {A, B}`
  - `loop([D]), isVisited = {A, B, C}`
  - `loop([]), isVisited = {A, B, C, D}`
- `loop([D]), isVisited = {A, B, C, D}` // we stop here but we already visited all members.

---

Now, we call `loop(tail)` **outside of if(isVisited(head)) block**, we keep searching in the first scenario 👍 👍 👍

Reply:
> So, previously, we stopped looping when the `head` is already visited. This is a problem.

Exactly, we're omitting `tail` even though it might contain unseen build targets. Thanks for putting detailed explanation @tanishiking